bindings/cs/MongoDB.Libmongocrypt/Callbacks/CipherCallbacks.cs

```diff
@@ -0,0 +1,130 @@
+﻿/*
+ * Copyright 2020–present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using System;
+using System.Security.Cryptography;
+
+namespace MongoDB.Libmongocrypt.Callbacks
+{
+    internal enum CryptMode
+    {
+        Encrypt,
+        Decrypt
+    }
+
+    internal static class CipherCallbacks
+    {
+        public static bool Encrypt(
+            IntPtr ctx,
+            IntPtr key,
+            IntPtr iv,
+            IntPtr @in,
+            IntPtr @out,
+            ref uint bytes_written,
+            IntPtr statusPtr)
+        {
+            var status = new Status(StatusSafeHandle.FromIntPtr(statusPtr));
+            try
+            {
+                var keyBinary = new Binary(BinarySafeHandle.FromIntPtr(key));
+                var inputBinary = new Binary(BinarySafeHandle.FromIntPtr(@in));
+                var outputBinary = new Binary(BinarySafeHandle.FromIntPtr(@out));
+                var ivBinary = new Binary(BinarySafeHandle.FromIntPtr(iv));
+
+                byte[] keyBytes = keyBinary.ToArray();
+                byte[] ivBytes = ivBinary.ToArray();
+                byte[] inputBytes = inputBinary.ToArray();
+
+                var outputBytes = AesCrypt(keyBytes, ivBytes, inputBytes, CryptMode.Encrypt);
+                bytes_written = (uint)outputBytes.Length;
```

Comment:
@kevinAlbs why you cannot get the size of `bytes_written` from the `outputBytes`. Why do we need to specify this value on the level of the binding?

Reply:
is it because the allocated memory for `out` can be bigger than `outputBytes` and you cannot determine(for some reasons) the exact arrays size on the `c` level?

Reply:
It is a little redundant. The expected number of bytes written could be determined in libmongocrypt. But currently it is used as an extra safety check, to ensure that the bindings write an output that is divisible by the block size: https://github.com/mongodb/libmongocrypt/blob/752fc0b6d907561c51e2671ad1bef4936d92d724/src/mongocrypt-crypto.c#L343